crates/biome_js_analyze/src/analyzers/style/use_consistent_array_type.rs

```diff
@@ -0,0 +1,485 @@
+use biome_analyze::{
+    context::RuleContext, declare_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+};
+use biome_console::{markup, Markup, MarkupBuf};
+use biome_diagnostics::Applicability;
+use biome_js_factory::make;
+use biome_js_syntax::{
+    AnyTsName, AnyTsType, JsSyntaxKind, JsSyntaxToken, TriviaPieceKind, TsReferenceType,
+    TsTypeAnnotation, TsTypeArguments, T,
+};
+use biome_rowan::{declare_node_union, AstNode, AstSeparatedList, BatchMutationExt, TriviaPiece};
+use serde::{Deserialize, Serialize};
+
+use crate::JsRuleAction;
+
+declare_rule! {
+    /// Require consistently using either `T[]` or `Array<T>`
+    ///
+    /// ESLint (typescript-eslint) equivalent: [array-type](https://typescript-eslint.io/rules/array-type)
+    /// ## Example
+    ///
+    /// ### Invalid
+    /// ```ts,expect_diagnostic
+    /// let invalid: Array<foo>;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// let invalid: Promise<Array<string>>;
+    /// ```
+    ///
+    /// ```ts,expect_diagnostic
+    /// ```
+    ///
+    /// ### Valid
+    /// ```ts
+    /// const valid: Array<string | number> = ['a', 'b'];
+    /// const valid: Array<{ prop: string }> = [{ prop: 'a' }];
+    /// const valid: Array<() => void> = [() => {}];
+    /// const valid: MyType[] = ['a', 'b'];
+    /// const valid: string[] = ['a', 'b'];
+    /// const valid: readonly string[] = ['a', 'b'];
+    /// ```
+    ///
+    /// ## Options
+    /// The rule provides two options that help to specify what type of array declarations to use.
+    ///
+    /// Default: "simple"
+    ///
+    /// ```json
+    /// {
+    ///     "//": "...",
+    ///     "options": {
+    ///         "consistentArrayType": "simple" | "generic"
+    ///     },
+    /// }
+    /// ```
+    ///
+    /// By default, all array declarations will be converted to `T[]` or `readonly T[]`, which it means `simple`,
+    /// or if the options is set to the "generic", that all will converted to `Array<T>` or `ReadonlyArray<T>`.
+    ///
+    pub(crate) UseConsistentArrayType {
+        version: "1.4.2",
+        name: "useConsistentArrayType",
+        recommended: false,
+        fix_kind: FixKind::Unsafe,
+    }
+}
+
+declare_node_union! {
+    pub(crate) Query = TsTypeAnnotation | TsReferenceType
+}
+
+#[derive(Debug, Copy, Clone)]
+enum TsArrayKind {
+    /// `T[]`
+    Simple,
+    /// `Array<T>`
+    GenericArray,
+    /// `readonly T[]`
+    Readonly,
+    /// `ReadonlyArray<T>`
+    ReadonlyGenericArray,
+}
+
+impl Rule for UseConsistentArrayType {
+    type Query = Ast<Query>;
```

Comment:
I think we could simplify the implementation by directly querying `AnyTsType`.
```suggestion
    type Query = Ast<AnyTsType>;
```

Reply:
If used `AnyTsType` here, when parsing this code `let invalid: readonly number[];`

```rs
  AnyTsType::TsArrayType(_) | AnyTsType::TsTypeOperatorType(_) => {
          if options.consistent_array_type == "simple" {
              return None;
          }
          match get_array_kind_by_any_type(query) {
              Some(array_kind) => transform_array_type(ty, array_kind),
              _ => None,
          }
   }
```
And then the above code will hit twice, the first is `TypeOperatorType`:
```
> 1 │  let invalid: readonly number[];
      │               ^^^^^^^^^^^^^^^^^
```
 and the second is `TsArrayType`:
```
  > 1 │  let invalid: readonly number[];
      │                        ^^^^^^^^
```

I think it's a duplicate.

Reply:
Oh I see.
One way to address this could be ensuring that the parent is not itself a type operator:

```rs
AnyTsType::TsArrayType(_)
| AnyTsType::TsTypeOperatorType(_) if query.parent().kind() != Some(JsSyntaxKind::TS_TYPE_OPEARTOR) => {
}

```

Reply:
Done~